### PR TITLE
Feature/cats instances

### PR DESCRIPTION
**Pull request by vaslabs**
Following #1087 

Provides Invariant and Semigroupal cats instances for Lens.

## Motivation
Get some out of the box syntactic sugar from cats

I added a few Laws tests for Lens and for Getter (noticed it had Arrow already).

### Intersecting lenses

I've added a test to demonstrate a potential UX(?) issue discussed with @julien-truffaut 

```scala
test("tupled breaks get what you replace if you tuple the same/conflicting lenses") {
    val zippedLenses: Lens[Point, (Int, Int)] = (x, x).tupled
    assertEquals(
      zippedLenses.replace((2, 3))(Point(5, 6)),
      Point(3, 6)
    ) //2 disappeared

  }
  ```

Should I drop it in a doc somewhere?

### Laws tests
The tests rely on a generator method provider from cats to sample test the Eq instance (because we cannot prove two Lens instances are equal until we generate equal objects that they operate on. So this generates some tuple `S -> A` , creates a Lens on `S` and sets `A`, and yields (via via `lens.replace` and `lens.get` tuple) a new `S -> A` to be derived Equal from an `Eq[(S, A)]` instance.

Bit of coding gymnastics so I'd be a bit sceptical maintaining this, especially since sampleEq is deprecated.

Please let me know your thoughts.

### CODE REVIEW

File: test/shared/src/test/scala/monocle/LensInstancesSpec.scala [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+  implicit val arbSampleLensBool: Arbitrary[Lens[Sample, Boolean]] = Arbitrary(
+    Gen.const(Lens[Sample, Boolean](_.c)(b => s => s.copy(c = b)))
+  )
+
+  checkAll("Lens.InvariantLaws", InvariantTests[Lens[Sample, *]].invariant[Char, Int, Boolean])
+
+  checkAll("Lens.SemigroupalLaws", SemigroupalTests[Lens[Sample, *]].semigroupal[Char, Int, Boolean])
+
+  //Explicit tests for clarity
+  test("conflicting lenses are left associative") {
```
Comment (julien-truffaut) [inline test/shared/src/test/scala/monocle/LensInstancesSpec.scala:37]: what do you mean by left associative?

**Comment by julien-truffaut** [comment]
Thanks @vaslabs 
The cats spec seems to be a pain to setup, not sure if it is worth it. Maybe we could only zip two lenses (with different fields) and run the `LensTests`. 

It is good idea to have a unit test to show where it could break the Lens property.

File: test/shared/src/test/scala/monocle/LensInstancesSpec.scala (hunk 1 above)
Comment (vaslabs) [reply test/shared/src/test/scala/monocle/LensInstancesSpec.scala:37]: maybe wrong language, I meant the right setter operation overwrites the left one, so if they conflict the right most operation (lens replace) result is the result we get

**Comment by vaslabs** [comment]
> Thanks @vaslabs
> The cats spec seems to be a pain to setup, not sure if it is worth it. Maybe we could only zip two lenses (with different fields) and run the `LensTests`.
> 
> It is good idea to have a unit test to show where it could break the Lens property.

I agree, only had it in initial PR as proof

File: test/shared/src/test/scala/monocle/LensInstancesSpec.scala (hunk 1 above)
Comment (vaslabs) [reply test/shared/src/test/scala/monocle/LensInstancesSpec.scala:37]: I might need to find something better here, the API signature does not point to what's going on
```
zippedLenses.replace(('b', 'c'))
```
Any suggestions?

Comment (vaslabs) [reply test/shared/src/test/scala/monocle/LensInstancesSpec.scala:37]: will probably make a better test with fold or foldMap

Comment (vaslabs) [reply test/shared/src/test/scala/monocle/LensInstancesSpec.scala]: I've cleaned it up, let me know what you think

**Review by julien-truffaut — APPROVED** [review]
LGTM, thanks for your work!

**Pull request merged**